drivers/gpio/gpio_hpmicro.c

```diff
@@ -0,0 +1,259 @@
+/*
+ * Copyright (c) 2022 hpmicro
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+
+
+#define DT_DRV_COMPAT hpmicro_hpm_gpio
+
+#include <errno.h>
+#include <device.h>
+#include <drivers/gpio.h>
+#include <drivers/pinctrl.h>
+#include <soc.h>
+
+#include <hpm_gpio_drv.h>
+
+#include "gpio_utils.h"
+
+struct gpio_hpm_config {
+	struct gpio_driver_config common;
+	GPIO_Type *gpio_base;
+	uint32_t  port_base;
+#ifdef CONFIG_PINCTRL
+	const struct pinctrl_dev_config *pincfg;
+#endif /* CONFIG_PINCTRL */
```

Comment:
why does a GPIO driver require pinctrl? can you please explain

Reply:
This chip io function is complex, where the pinmux peripheral controls the multiplexing of io functions, drive capability, speed, pull-up and pull-down, analogue or digital functions, etc. The other part is the gpio controller, who is only responsible for input and output and status reading as well as the interrupt controller.